trie/proof.go

```diff
@@ -454,96 +454,120 @@ func hasRightElement(node node, key []byte) bool {
 //
 // Except returning the error to indicate the proof is valid or not, the function will
 // also return a flag to indicate whether there exists more accounts/slots in the trie.
-func VerifyRangeProof(rootHash common.Hash, firstKey []byte, lastKey []byte, keys [][]byte, values [][]byte, proof ethdb.KeyValueReader) (error, bool) {
+//
+// TODO(karalabe): Reject the proof if it contains extra data.
+func VerifyRangeProof(rootHash common.Hash, firstKey []byte, lastKey []byte, keys [][]byte, values [][]byte, proof ethdb.KeyValueReader) (ethdb.KeyValueStore, *Trie, bool, error) {
 	if len(keys) != len(values) {
-		return fmt.Errorf("inconsistent proof data, keys: %d, values: %d", len(keys), len(values)), false
+		return nil, nil, false, fmt.Errorf("inconsistent proof data, keys: %d, values: %d", len(keys), len(values))
 	}
 	// Ensure the received batch is monotonic increasing.
 	for i := 0; i < len(keys)-1; i++ {
 		if bytes.Compare(keys[i], keys[i+1]) >= 0 {
-			return errors.New("range is not monotonically increasing"), false
+			return nil, nil, false, errors.New("range is not monotonically increasing")
 		}
 	}
```

Comment:
Would't it make sense to have this check include `firstKey` and `lastKey` aswell? Something like this: 
```
	var pKey = firstKey
	for i := 0; i < len(keys)-1; i++ {
		if bytes.Compare(pKey, keys[i]) >= 0 {
			return nil, nil, false, errors.New("range is not monotonically increasing")
		}
		pKey = keys[i]
	}
	if bytes.Compare(pKey, keys[len(keys)-1]) >= 0 {
		return nil, nil, false, errors.New("range is not monotonically increasing")
	}
```

Reply:
Or rather
```golang
	// Ensure the received batch is monotonic increasing.
	if firstKey != nil && len(keys) > 0 {
		if bytes.Compare(firstKey, keys[0]) > 0 {
			return nil, nil, false, errors.New("range is not monotonically increasing [0]")
		}
	}
	for i := 0; i < len(keys)-1; i++ {
		if bytes.Compare(keys[i], keys[i+1]) >= 0 {
			return nil, nil, false, errors.New("range is not monotonically increasing [1]")
		}
	}
	if lastKey != nil && len(keys) > 0 {
		if bytes.Compare(keys[len(keys)-1], lastKey) > 0 {
			return nil, nil, false, errors.New("range is not monotonically increasing [2]")
		}
	}
```

Reply:
For the snap implementation, this doesn't mater, since AFAICT the snap callers always only send in the `origin` and the last key as `firstKey` and `lastKey`.